wgpu-hal/src/gles/egl.rs

```diff
@@ -160,6 +160,7 @@ fn choose_config(
 ) -> Result<(egl::Config, bool), crate::InstanceError> {
     //TODO: EGL_SLOW_CONFIG
     let tiers = [
+        ("surfaceless", &[egl::SURFACE_TYPE, egl::PBUFFER_BIT][..]),
```

Comment:
Well, in that case it could not find any config for GLES backend. Weird thing, according to logs it doesn't even try `surfaceless` tier. I could not find explanation to that.
Ran this: `RUST_LOG=info WAYLAND_DISPLAY= DISPLAY= WGPU_BACKEND=gl cargo run --example capture`
L163 commented out:
```
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] Loading Wayland library to get the current display
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] Loading X11 library to get the current display
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] No windowing system present. Using surfaceless platform
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] Enabling EGL debug output
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] Display vendor "Mesa Project", version (1, 5)
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] 	EGL surface: +srgb
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] 	Trying native-render
[2022-01-01T06:10:00Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] 	Trying presentation
[2022-01-01T06:10:00Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-01T06:10:00Z INFO  wgpu_hal::gles::egl] 	Trying off-screen
[2022-01-01T06:10:00Z WARN  wgpu_hal::gles::egl] No config found!
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', wgpu/examples/capture/main.rs:36:6
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2022-01-01T06:10:00Z INFO  wgpu_core::hub] Dropping Global
```
With it present:
```
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] Loading Wayland library to get the current display
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] Loading X11 library to get the current display
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] No windowing system present. Using surfaceless platform
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] Enabling EGL debug output
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] Display vendor "Mesa Project", version (1, 5)
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	EGL surface: +srgb
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	Trying native-render
[2022-01-01T06:11:18Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	Trying presentation
[2022-01-01T06:11:18Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	Trying off-screen
[2022-01-01T06:11:18Z WARN  wgpu_hal::gles::egl] EGL says it can present to the window but not natively
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	EGL context: +debug
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	EGL context: +robust access
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] 	EGL context: +surfaceless
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] Max label length: 256
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::egl] Enabling GLES debug output
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::adapter] Vendor: AMD
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::adapter] Renderer: AMD RENOIR (DRM 3.42.0, 5.15.5-200.fc35.x86_64, LLVM 13.0.0)
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::adapter] Version: OpenGL ES 3.2 Mesa 21.3.2
[2022-01-01T06:11:18Z INFO  wgpu_hal::gles::adapter] SL version: OpenGL ES GLSL ES 3.20
[2022-01-01T06:11:18Z INFO  wgpu_core::instance] Adapter Gl AdapterInfo { name: "AMD RENOIR (DRM 3.42.0, 5.15.5-200.fc35.x86_64, LLVM 13.0.0)", vendor: 4098, device: 0, device_type: DiscreteGpu, backend: Gl }
[2022-01-01T06:11:19Z INFO  wgpu_core::device] Created buffer Valid((0, 1, Gl)) with BufferDescriptor { label: None, size: 102400, usage: MAP_READ | COPY_DST, mapped_at_creation: false }
[2022-01-01T06:11:19Z INFO  wgpu_core::device] Created texture Valid((0, 1, Gl)) with TextureDescriptor { label: None, size: Extent3d { width: 100, height: 200, depth_or_array_layers: 1 }, mip_level_count: 1, sample_count: 1, dimension: D2, format: Rgba8UnormSrgb, usage: COPY_SRC | RENDER_ATTACHMENT }
[2022-01-01T06:11:19Z INFO  wgpu_core::device] Buffer (0, 1, Gl) is dropped
[2022-01-01T06:11:19Z INFO  wgpu_core::hub] Dropping Global
[2022-01-01T06:11:19Z INFO  wgpu_core::device] Destroying 1 command encoders
```

Reply:
That's quite silly, isn't it? The driver is happy to find a context if we request `SURFACE_TYPE = PBUFFER` but isn't happy if we are omitting this requirement. GL initialization is such a mess...
Note that in your changed scenario it accumulates all of the requirements from lower tiers.

it might help to just insert something like `dbg!(attributes)` before each `choose_first_config` call to see if anything is abnormal there.
Ultimately, I expect an empty list of attributes (with just `egl::NONE`) to succeed, but this would be good to confirm/check as well.

Reply:
@kvark You're absolutely right, it's getting accumulated...

With `surfaceless` tier:

<details>

```
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] Loading Wayland library to get the current display
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] Loading X11 library to get the current display
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] No windowing system present. Using surfaceless platform
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] Enabling EGL debug output
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] Display vendor "Mesa Project", version (1, 5)
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	EGL surface: +srgb
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	Trying native-render
[wgpu-hal/src/gles/egl.rs:190] &attributes = [
    12339,
    1,
    12352,
    4,
    12339,
    4,
    12333,
    1,
    12321,
    8,
    12344,
]
[2022-01-03T19:05:14Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	Trying presentation
[wgpu-hal/src/gles/egl.rs:190] &attributes = [
    12339,
    1,
    12352,
    4,
    12339,
    4,
    12321,
    8,
    12344,
]
[2022-01-03T19:05:14Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	Trying off-screen
[wgpu-hal/src/gles/egl.rs:190] &attributes = [
    12339,
    1,
    12352,
    4,
    12321,
    8,
    12344,
]
[2022-01-03T19:05:14Z WARN  wgpu_hal::gles::egl] EGL says it can present to the window but not natively
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	EGL context: +debug
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	EGL context: +robust access
[2022-01-03T19:05:14Z INFO  wgpu_hal::gles::egl] 	EGL context: +surfaceless

```

</details>

Commented out:

<details>

```
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] Loading Wayland library to get the current display
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] Loading X11 library to get the current display
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] No windowing system present. Using surfaceless platform
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] Enabling EGL debug output
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] Display vendor "Mesa Project", version (1, 5)
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] 	EGL surface: +srgb
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] 	Trying native-render
[wgpu-hal/src/gles/egl.rs:190] &attributes = [
    12352,
    4,
    12339,
    4,
    12333,
    1,
    12321,
    8,
    12344,
]
[2022-01-03T19:06:50Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] 	Trying presentation
[wgpu-hal/src/gles/egl.rs:190] &attributes = [
    12352,
    4,
    12339,
    4,
    12321,
    8,
    12344,
]
[2022-01-03T19:06:50Z WARN  wgpu_hal::gles::egl] No config found!
[2022-01-03T19:06:50Z INFO  wgpu_hal::gles::egl] 	Trying off-screen
[wgpu-hal/src/gles/egl.rs:190] &attributes = [
    12352,
    4,
    12321,
    8,
    12344,
]
[2022-01-03T19:06:50Z WARN  wgpu_hal::gles::egl] No config found!
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', wgpu/examples/capture/main.rs:36:6
```

</details>

Reply:
I guess we can try just adding your SURFACE_TYPE attribute to the existing "off-screen" instead of adding a new tier.
It's crappy to provide 2 different surface type attributes, so I wonder how this would go.

Reply:
Done that

Reply:
Thanks! That seems to work. At least Intel isn't confused by multiple surface types.